packages/notifier/src/publish-kit.js

```diff
@@ -180,12 +200,221 @@ export const makePublishKit = () => {
       advanceCurrent(true, finalValue);
     },
     fail: reason => {
-      const rejection = HandledPromise.reject(reason);
-      // Suppress unhandled rejection error.
-      rejection.catch(() => {});
-      advanceCurrent(true, undefined, rejection);
+      advanceCurrent(true, undefined, makeQuietRejection(reason));
     },
   });
   return harden({ publisher, subscriber });
 };
 harden(makePublishKit);
+
+// TODO: Move durable publish kit to a new file?
+
+/**
+ * @typedef {ReturnType<typeof initDurablePublishKitState>} DurablePublishKitState
+ */
+
+/**
+ * @param {DurablePublishKitOptions} [options]
+ */
+const initDurablePublishKitState = (options = {}) => {
+  const { valueDurability = 'mandatory' } = options;
+  assert.equal(valueDurability, 'mandatory');
+  return {
+    // configuration
+    valueDurability,
+
+    // lifecycle progress
+    publishCount: 0n,
+    status: 'live', // | 'finished' | 'failed'
+
+    // persisted result data
+    hasValue: false,
+    value: undefined,
+  };
+};
+
+/** @type {WeakMap<object, {currentP, tailP, tailR}>} */
+const durablePublishKitEphemeralData = new WeakMap();
+
+// XXX Does state actually work as a WeakMap key?
```

Comment:
Likely no. Every time the same virtual object is paged in, a fresh representative is created with a fresh state record. I think we made the exposed WeakMaps "safe" for using virtual object representatives as weakmap keys. But perhaps not the state record. In any case, even if it works, it would be "wrong" (heavy air quotes) because the representative is intended to represent a long lived (virtual) object that can be accessed from outside the object. The state record is only designed to be accessed by the object (or object kit) for which it is the state record.

Could you use a virtual object instead, rather than its state record?

@FUDCo @warner , how close did I get? What's the real truth?

Reply:
When I try to use `this.facets` rather than `this.state` I get "Invalid value used as weak map key" errors, so I'm not sure how to key the ephemeral data. Adding a unique dummy object to the state of each durable publish kit would presumably suffer the same problem as using state directly... I could add an instance counter to baggage in `vivifyDurablePublishKit` and use that as the key to a _strong_ map associated with the baggage, but that seems wrong in its own way. I welcome some guidance here!

Reply:
@erights is correct. While using `state` as a WeakMap key cannot cause non-determinism or be used as a GC sensor (due to all the work we put into the [VirtualObjectAwareWeakMap](https://github.com/Agoric/agoric-sdk/blob/c7d422343c9fdfd173b6e756ad2a02577d7c4574/packages/SwingSet/src/liveslots/virtualObjectManager.js#L284)), it's not meant to be used that way. Using the `Publisher` facet would be more "with the grain".

I'm not sure why `this.facets` failed; unless the new class/`this`-based API means `this.facets` should be spelled `this.context.facets` or something, I'd expect it to behave basically the same as `this.state`. But neither `state` nor `facets` are meant to be used for their identity: only the facets themselves are made for that.

I think the immediate consequence of using `state` is that objects will be kept around longer than strictly necessary (i.e. increased RAM usage). However I'm not sure how much of that we can avoid. So perhaps we should land this as-is, and then continue to pursue further optimization later.

(BTW my [original sketch](https://github.com/Agoric/agoric-sdk/issues/4567#issuecomment-1088182184) for this did indeed use an integer counter as the key to a strong Map, as @gibson042 describes, and re-reading the discussion there reminds me that I didn't think we could build a high-cardinality yet low-RAM version back in April either, so yeah, trying to make this RAM efficient might not be the highest priority)

## Further Details

The `state` object is just another anonymous JS object: unlike the facets (i.e. Representatives) it lacks a `vref` identity. The same is true for the other objects we create in the course of making the Representative, like `facets` (the `{ publisher, subscriber }` container which holds them), and the `context` record which holds `{ facets, state }`. However we link these four kinds of objects (`state`, `facets`, `context`, and the facets themselves) to make sure that userspace cannot sense GC by e.g. holding onto `state`, waiting for the facets to get dropped and collected (or not), pulling `publisher` back out of a virtual collection, then comparing the new `state` object against the held copy: if they're the same, GC did not happen, if they differ, GC did happen. `context` contains `facets`, and `facets` contains the facets, and each facet closes over a reference to `state`. Then, by [maintaining an edge](https://github.com/Agoric/agoric-sdk/blob/c7d422343c9fdfd173b6e756ad2a02577d7c4574/packages/SwingSet/src/liveslots/virtualObjectManager.js#L752) from `state` to the [cohort of facets](https://github.com/Agoric/agoric-sdk/blob/c7d422343c9fdfd173b6e756ad2a02577d7c4574/packages/SwingSet/src/liveslots/virtualObjectManager.js#L742), we form a cycle, so `state` and `facets` and the facets all keep each other alive.

In addition, the `state` object is [marked as "unweakable"](https://github.com/Agoric/agoric-sdk/blob/c7d422343c9fdfd173b6e756ad2a02577d7c4574/packages/SwingSet/src/liveslots/virtualObjectManager.js#L718), so when it is used as a WeakMap/WeakSet key, we add an extra reference edge from the WeakMap to the `state`, to ensure that `state` sticks around at least as long as the WeakMap. We do the same for `context`. This prevents userspace from building a recognizer for one of these objects with `weakset.add()`, then probing it with a later copy to see if they differ, revealing that GC happened. By forcing `state` to be held strongly, we inhibit the GC, so the recognizer will always give the same answer.

The net result is that these internal objects cannot be used to sense GC, but using them as a WeakMap key will actually cause them to be held strongly, until either the WeakMap itself goes away, or someone explicitly removes it.

Getting back to publish-kit.js, with the current PR, the `publisher` and `subscriber` facets will be kept in RAM whenever `state` is being used as a key in the `durablePublishKitEphemeralData` WeakMap, and that happens the first time any method is called on either facet (and is never removed). So in practice, every cohort will be held in RAM all the time, along with all of its state, making this consume just as much RAM as a non-virtual version. They will be durable, but high cardinality will be a RAM problem.

Using `state` or `facets` or `context` as the WeakMap key doesn't let you take advantage of virtualization, because none of them have a vref identity. Things could be better if you use one of the *facets* as the WeakMap key, like `publisher`, or a third (empty) facet that you create just to serve as a key. In this case, the VirtualObjectAwareWeakMap [uses the facet's vref](https://github.com/Agoric/agoric-sdk/blob/c7d422343c9fdfd173b6e756ad2a02577d7c4574/packages/SwingSet/src/liveslots/virtualObjectManager.js#L320) (a string) as the key of a strong Map, pointing to the value. It allows the facet (the Representative) to be dropped (from RAM), which allows the `state` data to be dropped (from RAM), along with the rest of the cohort. They will be paged back in if/when someone invokes a method on either facet, but in the meantime, we only hold a vref string in RAM, not the whole `{ publisher, subscriber }` cohort. `WeakMap` is not virtual, it's just *aware* of virtual objects.

Doing that would save you the RAM of the facets, however the design is still keeping the full resolution value of all publishers in RAM. To reduce usage further, we want to move that to DB. It looks like `durablePublishKitEphemeralData` is used to hold two promises (`currentP` and `tailP`), and one bare resolver function (`tailR`). `currentP` may be resolved to some arbitrary data (the most recently published value), or it might be unresolved. I see that `value` is stored on `state.value` (thus in the durable object, thus in the DB, good), but it looks like we're also holding a copy of `value` in a resolved Promise, which means holding the resolution data in RAM (bad). So this approach will keep all the current resolution values in RAM, even though we don't need them until a new subscriber comes along.

A better approach would retain fewer Promises. I think the two dominant cases are a new subscriber to a non-empty Publisher (i.e. they should get the current value), or a subscriber who is up-to-date and will be waiting for a value. In the first case, I think we could return a new Promise (freshly resolved to the current value) to each new subscriber, rather than keeping around a pre-resolved Promise for them. (Note that promise ID `vpid`s are retired at the vat/kernel boundary upon resolution, so you don't save any syscalls by sending the same resolved promise over the wire multiple times: each one requires a new `vpid` allocation and a new `syscall.resolve()`). The current value would remain in the DB, in `state.value`, until needed.

For the second case, I think we *do* need to hold onto a single Promise in RAM, but it should be unresolved. We need the WeakMap or a BigScalarMapStore to map from something with a vref (the `publisher` facet) to a pair of the unresolved Promise and a resolver function. The RAM usage here will be O(N) in the number of publishers with at least one active subscriber, but unaffected by the total amount of resolution data, or the number of unloved publishers, all of which will only consume DB space.

Reply:
I definitely want to discuss this further face-to-face, but in the meantime I've just taken the primary suggestion to key the ephemeral-data WeakMap with a facet. Assuming CI looks good, this should be clear to merge.